core/services/orm/tree_traversal/RelationNode.php

```diff
@@ -0,0 +1,214 @@
+<?php
+
+namespace EventEspresso\core\services\orm\tree_traversal;
+
+use EE_Base_Class;
+use EE_Error;
+use EE_Model_Relation_Base;
+use EEM_Base;
+use EventEspresso\core\exceptions\InvalidDataTypeException;
+use EventEspresso\core\exceptions\InvalidInterfaceException;
+use EventEspresso\core\services\payment_methods\forms\PayPalSettingsForm;
+use InvalidArgumentException;
+use ReflectionException;
+
+/**
+ * Class RelationNode
+ *
+ * Wraps a model object and one of its model's relations; stores how many related model objects exist across that
+ * relation, and eventually createsa a ModelObjNode for each of its related model objects.
+ *
+ * @package     Event Espresso
+ * @author         Mike Nelson
+ * @since         $VID:$
+ *
+ */
+class RelationNode extends BaseNode
+{
+    /**
+     * @var EE_Base_Class
+     */
+    protected $main_model_obj;
+
+    /**
+     * @var int
+     */
+    protected $count;
+
+    /**
+     * @var EEM_Base
+     */
+    protected $related_model;
+
+
+    protected $model_obj_nodes;
+
+    public function __construct($main_model_obj, $related_model)
+    {
+        $this->main_model_obj = $main_model_obj;
+        $this->related_model = $related_model;
+        $this->model_obj_nodes = [];
+    }
+
+
+    /**
+     * Here is where most of the work happens. We've counted how many related model objects exist, here we identify
+     * them (ie, learn their IDs). But its recursive, so we'll also find their related dependent model objects etc.
+     * @since $VID:$
+     * @param int $model_objects_to_identify
+     * @return int
+     * @throws EE_Error
+     * @throws InvalidArgumentException
+     * @throws InvalidDataTypeException
+     * @throws InvalidInterfaceException
+     * @throws ReflectionException
+     */
+    protected function work($model_objects_to_identify)
+    {
+        $num_identified = $this->visitAlreadyDiscoveredNodes($this->model_obj_nodes, $model_objects_to_identify);
+        if ($num_identified < $model_objects_to_identify) {
+            $related_model_objs = $this->related_model->get_all(
+                [
+                    $this->whereQueryParams(),
+                    'limit' => [
+                        count($this->model_obj_nodes),
+                        $model_objects_to_identify
+                    ]
+                ]
+            );
+            $new_item_nodes = [];
+
+            // Add entity nodes for each of the model objects we fetched.
+            foreach ($related_model_objs as $related_model_obj) {
+                $entity_node = new ModelObjNode($related_model_obj);
+                $this->model_obj_nodes[ $related_model_obj->ID() ] = $entity_node;
```

Comment:
using SplDoublyLinkedList this would be something like:

```php
$entity_node = new ModelObjNode($related_model_obj);
$this->add( $related_model_obj->ID(), $entity_node );
```

again, just a suggestion, that I think would help

Reply:
🙇